File: app/src/main/java/com/jon/cotgenerator/ui/EditPresetActivity.java
package com.jon.cotgenerator.ui;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.preference.PreferenceManager;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.jon.cotgenerator.R;
import com.jon.cotgenerator.presets.OutputPreset;
import com.jon.cotgenerator.presets.PresetRepository;
import com.jon.cotgenerator.service.CotService;
import com.jon.cotgenerator.utils.InputValidator;
import com.jon.cotgenerator.utils.Key;
import com.jon.cotgenerator.utils.Notify;
import com.jon.cotgenerator.utils.PrefUtils;
import com.jon.cotgenerator.utils.Protocol;

public class EditPresetActivity
        extends ServiceBoundActivity {

    private SharedPreferences prefs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(R.string.toolbarHeaderEditPreset);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        Fragment fragment = EditPresetFragment.newInstance();
        /* Pass all inter-activity arguments directly to the fragment */
        fragment.setArguments(getIntent().getExtras());
        getSupportFragmentManager().beginTransaction()
                .replace(R.id.fragment, fragment)
                .commit();

        prefs = PreferenceManager.getDefaultSharedPreferences(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.edit_preset_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
            case R.id.save:
                if (settingsAreValid()) {
                    storePresetInDatabase();
                    passPresetBackToMainActivity();
                }
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        new MaterialAlertDialogBuilder(this)
                .setTitle(R.string.goBack)
                .setMessage(R.string.goBackMessage)
                .setPositiveButton(android.R.string.ok, (dialog, i) -> super.onBackPressed())
                .setNegativeButton(android.R.string.cancel, (dialog, i) -> dialog.dismiss())
                .show();
    }

    private boolean settingsAreValid() {
        try {
            /* Regular options */
            String protocolStr = PrefUtils.getString(prefs, Key.PRESET_PROTOCOL);
            String alias = PrefUtils.getString(prefs, Key.PRESET_ALIAS);
            String address = PrefUtils.getString(prefs, Key.PRESET_DESTINATION_ADDRESS);
            String portStr = PrefUtils.getString(prefs, Key.PRESET_DESTINATION_PORT);
            if (!InputValidator.validateString(protocolStr, ".*?")) {
                Notify.orange(getRootView(), "No protocol chosen!");
                return false;
            } else if (!InputValidator.validateString(alias, ".*?")) {
                Notify.orange(getRootView(), "Enter an alias to describe the preset!");
                return false;
            } else if (!InputValidator.validateString(address, ".*?")) {
                Notify.orange(getRootView(), "No destination address has been entered!");
                return false;
            } else if (!InputValidator.validateInt(portStr, 1, 65535)) {
                Notify.orange(getRootView(), "Invalid port number!");
                return false;
            }

            Protocol protocol = getInputProtocol();
            if (protocol == Protocol.SSL) {
                /* SSL-only options */
                String clientBytes = PrefUtils.getString(prefs, Key.PRESET_SSL_CLIENTCERT_BYTES);
                String clientPass = PrefUtils.getString(prefs, Key.PRESET_SSL_CLIENTCERT_PASSWORD);
                String trustBytes = PrefUtils.getString(prefs, Key.PRESET_SSL_TRUSTSTORE_BYTES);
                String trustPass = PrefUtils.getString(prefs, Key.PRESET_SSL_TRUSTSTORE_PASSWORD);
                if (!InputValidator.validateString(clientBytes)) {
                    Notify.orange(getRootView(), "No client certificate chosen!");
                    return false;
                } else if (!InputValidator.validateString(clientPass)) {
                    Notify.orange(getRootView(), "No client certificate password entered!");
                    return false;
                } else if (!InputValidator.validateString(trustBytes)) {
                    Notify.orange(getRootView(), "No trust store chosen!");
                    return false;
                } else if (!InputValidator.validateString(trustPass)) {
                    Notify.orange(getRootView(), "No trust store password entered!");
                    return false;
                }
            }
        } catch (Exception e) {
            Notify.orange(getRootView(), "Unknown error when validating preset: " + e.getMessage());
            return false;
        }
        return true;
    }

    private void storePresetInDatabase() {
        Protocol protocol = getInputProtocol();
        OutputPreset preset = new OutputPreset(
                protocol,
                PrefUtils.getString(prefs, Key.PRESET_ALIAS),
                PrefUtils.getString(prefs, Key.PRESET_DESTINATION_ADDRESS),
                Integer.parseInt(PrefUtils.getString(prefs, Key.PRESET_DESTINATION_PORT))
        );
        if (protocol == Protocol.SSL) {
            preset.clientCert = PrefUtils.getString(prefs, Key.PRESET_SSL_CLIENTCERT_BYTES).getBytes();
            preset.trustStore = PrefUtils.getString(prefs, Key.PRESET_SSL_TRUSTSTORE_BYTES).getBytes();
            preset.clientCertPassword = PrefUtils.getString(prefs, Key.PRESET_SSL_CLIENTCERT_PASSWORD);
            preset.trustStorePassword = PrefUtils.getString(prefs, Key.PRESET_SSL_TRUSTSTORE_PASSWORD);
        }
        PresetRepository repository = PresetRepository.getInstance();
        repository.insertPreset(preset);
    }

    private void passPresetBackToMainActivity() {
        OutputPreset preset = new OutputPreset(
                getInputProtocol(),
                PrefUtils.getString(prefs, Key.PRESET_ALIAS),
                PrefUtils.getString(prefs, Key.PRESET_DESTINATION_ADDRESS),
                Integer.parseInt(PrefUtils.getString(prefs, Key.PRESET_DESTINATION_PORT))
        );
        Intent intent = new Intent();
        intent.setData(Uri.parse(preset.toString()));
        setResult(RESULT_OK, intent);
        finish();
    }

    private View getRootView() {
        return findViewById(android.R.id.content);
    }

    private Protocol getInputProtocol() {
        return Protocol.fromString(PrefUtils.getString(prefs, Key.PRESET_PROTOCOL));
    }

    @Override
    public void onStateChanged(CotService.State state, @Nullable Throwable throwable) {
        if (state == CotService.State.ERROR && throwable != null) {
            Notify.red(getRootView(), "Error: " + throwable.getMessage());
        }
    }
}

File: common/src/main/java/com/jon/common/ui/editpreset/EditPresetActivity.kt
package com.jon.common.ui.editpreset

import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.preference.PreferenceManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.jon.common.R
import com.jon.common.prefs.CommonPrefs
import com.jon.common.prefs.getStringFromPair
import com.jon.common.prefs.parseIntFromPair
import com.jon.common.presets.OutputPreset
import com.jon.common.repositories.PresetRepository
import com.jon.common.ui.ServiceBoundActivity
import com.jon.common.utils.InputValidator
import com.jon.common.utils.Notify
import com.jon.common.utils.Protocol

class EditPresetActivity : ServiceBoundActivity() {

    private val presetRepository = PresetRepository.getInstance()
    private val prefs: SharedPreferences by lazy { PreferenceManager.getDefaultSharedPreferences(this) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity)
        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar?.let {
            it.setTitle(R.string.toolbar_header_edit_preset)
            it.setDisplayHomeAsUpEnabled(true)
        }

        /* Pass all inter-activity arguments directly to the fragment */
        val fragment = EditPresetFragment()
        fragment.arguments = intent.extras
        supportFragmentManager.beginTransaction()
                .replace(R.id.fragment, fragment)
                .commit()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.edit_preset_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val resId = item.itemId
        if (resId == android.R.id.home) {
            onBackPressed()
        } else if (resId == R.id.save) {
            if (settingsAreValid()) {
                if (EditPresetFragment.initialPresetValues == null) {
                    /* It's a new preset, so insert it into the database */
                    storePresetInDatabase()
                } else {
                    /* It's an update of an existing preset, so overwrite its DB record properly */
                    overwritePresetInDatabase()
                }
                passPresetBackToMainActivity()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        MaterialAlertDialogBuilder(this)
                .setTitle(R.string.go_back)
                .setMessage(R.string.go_back_message)
                .setPositiveButton(android.R.string.ok) { _, _ -> super.onBackPressed() }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
    }

    private fun settingsAreValid(): Boolean {
        val inputValidator = InputValidator()
        try {
            /* Regular options */
            val protocolStr: String = prefs.getStringFromPair(CommonPrefs.PRESET_PROTOCOL)
            val alias: String = prefs.getStringFromPair(CommonPrefs.PRESET_ALIAS)
            val address: String = prefs.getStringFromPair(CommonPrefs.PRESET_DESTINATION_ADDRESS)
            val portStr: String = prefs.getStringFromPair(CommonPrefs.PRESET_DESTINATION_PORT)
            if (!inputValidator.validateString(protocolStr, ".*?")) {
                Notify.orange(getRootView(), "No protocol chosen!")
                return false
            } else if (!inputValidator.validateString(alias, ".*?")) {
                Notify.orange(getRootView(), "Enter an alias to describe the preset!")
                return false
            } else if (!inputValidator.validateString(address, ".*?")) {
                Notify.orange(getRootView(), "No destination address has been entered!")
                return false
            } else if (!inputValidator.validateInt(portStr, 1, 65535)) {
                Notify.orange(getRootView(), "Invalid port number!")
                return false
            }
            if (getInputProtocol() == Protocol.SSL) {
                /* SSL-only options */
                val clientBytes: String = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_CLIENTCERT_BYTES)
                val clientPass: String = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_CLIENTCERT_PASSWORD)
                val trustBytes: String = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_TRUSTSTORE_BYTES)
                val trustPass: String = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_TRUSTSTORE_PASSWORD)
                if (!inputValidator.validateString(clientBytes)) {
                    Notify.orange(getRootView(), "No client certificate chosen!")
                    return false
                } else if (!inputValidator.validateString(clientPass)) {
                    Notify.orange(getRootView(), "No client certificate password entered!")
                    return false
                } else if (!inputValidator.validateString(trustBytes)) {
                    Notify.orange(getRootView(), "No trust store chosen!")
                    return false
                } else if (!inputValidator.validateString(trustPass)) {
                    Notify.orange(getRootView(), "No trust store password entered!")
                    return false
                }
            }
        } catch (e: Exception) {
            Notify.orange(getRootView(), "Unknown error when validating preset: " + e.message)
            return false
        }
        return true
    }

    private fun getEnteredPresetValues(): OutputPreset {
        val protocol = getInputProtocol()
        val preset = OutputPreset(
                protocol,
                prefs.getStringFromPair(CommonPrefs.PRESET_ALIAS),
                prefs.getStringFromPair(CommonPrefs.PRESET_DESTINATION_ADDRESS),
                prefs.getStringFromPair(CommonPrefs.PRESET_DESTINATION_PORT).toInt()
        )
        if (protocol == Protocol.SSL) {
            preset.clientCert = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_CLIENTCERT_BYTES).toByteArray()
            preset.trustStore = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_TRUSTSTORE_BYTES).toByteArray()
            preset.clientCertPassword = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_CLIENTCERT_PASSWORD)
            preset.trustStorePassword = prefs.getStringFromPair(CommonPrefs.PRESET_SSL_TRUSTSTORE_PASSWORD)
        }
        return preset
    }

    private fun storePresetInDatabase() {
        presetRepository.insertPreset(getEnteredPresetValues())
    }

    private fun overwritePresetInDatabase() {
        val original = EditPresetFragment.initialPresetValues
        val updated = getEnteredPresetValues()
        if (original == null) {
            presetRepository.insertPreset(updated)
        } else {
            presetRepository.updatePreset(original, updated)
        }
    }

    private fun passPresetBackToMainActivity() {
        val preset = OutputPreset(
                getInputProtocol(),
                prefs.getStringFromPair(CommonPrefs.PRESET_ALIAS),
                prefs.getStringFromPair(CommonPrefs.PRESET_DESTINATION_ADDRESS),
                prefs.parseIntFromPair(CommonPrefs.PRESET_DESTINATION_PORT))
        val intent = Intent()
        intent.data = Uri.parse(preset.toString())
        setResult(RESULT_OK, intent)
        finish()
    }

    private fun getInputProtocol() = Protocol.fromString(prefs.getStringFromPair(CommonPrefs.PRESET_PROTOCOL))
}